### PR TITLE
[AIRFLOW-3249] unify (and fix) pushing result to xcom

### PR DESCRIPTION
**Pull request by marengaz**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://jira.apache.org/jira/browse/AIRFLOW-3249) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://jira.apache.org/jira/browse/AIRFLOW-3249
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:
see https://jira.apache.org/jira/browse/AIRFLOW-3249

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.
  - All the public functions and the classes in the PR contain docstrings that explain what it does

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

**Comment by marengaz** [comment]
please merge PR #4056 before merging this one. we need `BaseOperator.do_xcom_push`!

**Comment by marengaz** [comment]
@feng-tao - resolved the conflict. thanks!

**Comment by marengaz** [comment]
hey @kaxil - this is the pr i was talking about yday. just fixed a new conflict in updating.md

File: UPDATING.md [hunk 1]
```diff
@@ -183,6 +188,12 @@ session.add(admin)
 session.commit()
 ```
 
+## Airflow 1.10.1
```
Comment (kaxil) [inline UPDATING.md]: This change shouldn't be here. Probably needs a rebase on master.

Comment (marengaz) [reply UPDATING.md]: pretty sure i fixed it

**Comment by feng-tao** [comment]
@marengaz , please don't close and open pr that frequent. We have already got warning from ASF for this issue.

**Comment by marengaz** [comment]
> @marengaz , please don't close and open pr that frequent. We have already got warning from ASF for this issue.

sorry @feng-tao! i'm doing it because it's the easiest way to rerun the travis build, which keeps failing sporadically (not my fault, different issue each time). what's the best thing to do in this scenario?

**Comment by ashb** [comment]
You can set up travis to run on your fork by following instructions here https://github.com/apache/airflow/blob/master/CONTRIBUTING.md#testing-on-travis-ci

**Comment by marengaz** [comment]
@ashb / @kaxil - please can you rerun the failed module/merge?
tests pass here - https://travis-ci.com/marengaz/airflow/builds/99818224

File: airflow/operators/bash_operator.py [hunk 2]
```diff
@@ -38,9 +38,6 @@ class BashOperator(BaseOperator):
     :param bash_command: The command, set of commands or reference to a
         bash script (must be '.sh') to be executed. (templated)
     :type bash_command: str
-    :param xcom_push: If xcom_push is True, the last line written to stdout
-        will also be pushed to an XCom when the bash command completes.
```
Comment (ashb) [inline airflow/operators/bash_operator.py:46]: I think  the info here that just the last line is returned to XCom needs to be preserved somewhere? (Possibly just as a new paragraph on line 37, before the params)

File: airflow/operators/docker_operator.py [hunk 3]
```diff
@@ -238,9 +235,8 @@ def execute(self, context):
             if result['StatusCode'] != 0:
                 raise AirflowException('docker container failed: ' + repr(result))
 
-            if self.xcom_push_flag:
-                return self.cli.logs(container=self.container['Id']) \
-                    if self.xcom_all else str(line)
+            return self.cli.logs(container=self.container['Id']) \
+                if self.xcom_all else line.encode('utf-8')
```
Comment (ashb) [inline airflow/operators/docker_operator.py]: Hmmm. This is a somewhat expensive operation if the result is going to be thrown away at the next stage...

**Review by ashb — COMMENTED** [review]
One small doc fix, and a question.

**Comment by marengaz** [comment]
agree on your comments @ashb. [build succeeded here](https://travis-ci.com/marengaz/airflow/builds/100585413) please can you re-review / merge?

**Review by kaxil — APPROVED** [review]

**Review by ashb — APPROVED** [review]

**Comment by kaxil** [comment]
Thanks @marengaz :) for the PR

**Comment by Fokko** [comment]
Nice one! Thanks @marengaz

**Pull request merged**